components/region_cache_memory_engine/src/range_manager.rs

```diff
@@ -0,0 +1,211 @@
+    }
+
+    pub(crate) fn new_range(&mut self, range: CacheRange) {
+        if let Some(sibling) = self.ranges.keys().find(|r| r.is_sibling(&range)).cloned() {
+            let left_sib = sibling < range;
+            let mut sib_meta = self.ranges.remove(&sibling).unwrap();
+            sib_meta.ranges_unreadable.insert(range.clone());
+            let mut range = CacheRange::merge(sibling, range);
+
+            // if sibling found above is the left sibling of the range, there could be a
+            // right sibling
+            if left_sib {
+                if let Some(right_sibling) =
+                    self.ranges.keys().find(|r| r.is_sibling(&range)).cloned()
+                {
+                    let right_sib_meta = self.ranges.remove(&right_sibling).unwrap();
+                    sib_meta.merge_meta(right_sib_meta);
+                    range = CacheRange::merge(range, right_sibling);
+                }
+            }
+            self.ranges.insert(range, sib_meta);
+        } else {
+            let mut range_meta = RangeMeta::default();
+            range_meta.ranges_unreadable.insert(range.clone());
+            self.ranges.insert(range, range_meta);
+        }
+    }
+
+    pub fn set_range_readable(&mut self, range: &CacheRange, set_readable: bool) {
+        let range_key = self
+            .ranges
+            .keys()
+            .find(|&r| r.contains(range))
+            .unwrap()
+            .clone();
+        let meta = self.ranges.get_mut(&range_key).unwrap();
+        meta.set_range_readable(range, set_readable);
+    }
+
+    pub fn set_safe_ts(&mut self, range: &CacheRange, safe_ts: u64) -> bool {
+        if let Some(meta) = self.ranges.get_mut(range) {
+            if meta.safe_ts > safe_ts {
+                return false;
+            }
+            meta.safe_ts = safe_ts;
+            true
+        } else {
+            false
+        }
+    }
+
+    pub(crate) fn overlap_with_range(&self, range: &CacheRange) -> bool {
+        self.ranges.keys().any(|r| r.overlaps(range))
+    }
+
+    pub(crate) fn range_snapshot(&mut self, range: &CacheRange, read_ts: u64) -> bool {
+        let Some(range_key) = self.ranges.keys().find(|&r| r.contains(range)) else {
+            return false;
+        };
+        let meta = self.ranges.get(range_key).unwrap();
+
+        if read_ts <= meta.safe_ts {
+            // todo(SpadeA): add metrics for it
+            return false;
+        }
+
+        if meta.ranges_evcited.iter().any(|r| r.overlaps(range)) {
+            return false;
+        }
+
+        if meta.ranges_unreadable.iter().any(|r| r.overlaps(range)) {
+            return false;
+        }
+
+        let meta = self.ranges.get_mut(&range_key.clone()).unwrap();
+        meta.range_snapshot_list
+            .entry(range.clone())
+            .or_default()
+            .new_snapshot(read_ts);
+        true
+    }
+
+    pub(crate) fn remove_range_snapshot(&mut self, range: &CacheRange, read_ts: u64) {
```

Comment:
pub(crate) fn remove_range_snapshot(&mut self, range: &CacheRange, read_ts: u64) {
```suggestion
    pub(crate) fn remove_range_snapshot(&mut self,  snapshot &RangeSnapshotMeta) {
```